docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,342 @@
+
+### Read streaming data scenarios
+
+There are a couple of typical patterns that emerge when trying to read streaming data:
+
+* Given a stream of data, parse a single message.
+* Given a stream of data, parse all available messages.
+
+The following examples use the `TryParseMessage` method for parsing messages from a `ReadOnlySequence<byte>`. `TryParseMessage` parses a single message and update the input buffer to trim the parsed message from the buffer. `TryParseMessage` is not part of .NET, it's a user written method used in the following sections.
+
+```csharp
+bool TryParseMessage(ref ReadOnlySequence<byte> buffer, out Message message);
+```
+
+### Read a single message
+
+The following code reads a single message from a `PipeReader` and returns it to the caller.
+
+[!code-csharp[ReadSingleMsg](~/samples/snippets/csharp/pipelines/ReadSingleMsg.cs?name=snippet)]
+
+The preceding code:
+
+* Parses a single message.
+* Updates the consumed `SequencePosition` and examined `SequencePosition` to point to the start of the trimmed input buffer.
+
+The two `SequencePosition` arguments are updated because `TryParseMessage` removes the parsed message from the input buffer. Generally, when parsing a single message from the buffer, the examined position should be one of the following:
+
+* The end of the message.
+* The end of the received buffer if no message was found.
+
+The single message case has the most potential for errors. Passing the wrong values to *examined* can result in an out of memory exception or an infinite loop. For more information, see the [PipeReader common problems](#gotchas) section in this article.
+
+### Reading multiple messages
+
+The following code reads all messages from a `PipeReader` and calls `ProcessMessageAsync` on each.
+
+[!code-csharp[MyConnection1](~/samples/snippets/csharp/pipelines/MyConnection1.cs?name=snippet)]
+
+### Cancellation
+
+`PipeReader.ReadAsync`:
+
+* Supports passing a <xref:System.Threading.CancellationToken>.
+* Throws in an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.
+* Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` returns a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.
+
+[!code-csharp[MyConnection](~/samples/snippets/csharp/pipelines/MyConnection.cs?name=snippet)]
+
+<a name="gotchas"></a>
+
+### PipeReader common problems
+
+* Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed.
```

Comment:
```suggestion
* Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed to `examined` but not `consumed`.
```

I'm not sure quite what was intended by the second sentence. It might be better to remove the second sentence entirely.

Reply:
Yea it may make sense to remove it as this was written before I had the bad samples. Those show the cases more clearly